### DISCUSSION TITLE
[Option, Result, Error Handling] to change current Error Handling

### DISCUSSION THREAD

**Discussion by igotfr** [vlang/v#12740] [Ideas]
```v
fn e(x int) !int { // implicitly any Error
  if x < 5 {
    return 8
  }
  return error('greater than or equal 5')
}

fn n(x int) ?int { // implicitly None
  if x < 5 {
    return 8
  }
  return none
}

struct MyError1 {
  msg string
  code int
}

struct MyError2 {
  msg string
  code int
}

struct MyError3 {
  msg string
  code int
}

fn f(x int) (MyError1 | MyError3) ! int {
  if x < 5 {
    return 8
  } else if x < 10 {
    return MyError1{msg: 'Error'} // without IError()
  }
  return MyError3{msg: 'Invalid'}
}

println(f(4) or { println(err.msg) return }) // 8
println(f(9) or { println(err.msg) return }) // Error
println(f(12) or { println(err.msg) return }) // Invalid

fn g(x int) (MyError1 | MyError3) ? int {
// same as:
// fn g(x int) (MyError1 | MyError3 | None) ! int {
  if x < 5 {
    return 8
  } else if x  < 10 {
    return MyError1{msg: 'Error'}
  } else if x < 15 {
    return MyError3{msg: 'Invalid'}
  }
  return none
}

println(g(4) or { println(err.msg) return }) // 8
println(g(9) or { println(err.msg) return }) // Error
println(g(12) or { println(err.msg) return }) // Invalid
println(g(17) or { println(err) return }) // none

fn f_invalid(x int) (MyError1 | MyError3) ! int {
  if x < 5 {
    return 8
  } else if x  < 10 {
    return MyError1{msg: 'Error'}
  } else if x < 15 {
    return MyError2{msg: 'Invalid'}
  }
  return MyError3{msg: 'Failure'}
}

println(f_invalid(4) or { println(err.msg) return }) // `compile error`
println(f_invalid(9) or { println(err.msg) return }) // `compile error`
println(f_invalid(12) or { println(err.msg) return }) // `compile error`
println(f_invalid(17) or { println(err) return }) // `compile error`

fn g_invalid(x int) (MyError1 | MyError3) ! int {
  if x < 5 {
    return 8
  } else if x  < 10 {
    return MyError1{msg: 'Error'}
  } else if x < 15 {
    return MyError3{msg: 'Invalid'}
  }
  return none
}

println(g_invalid(4) or { println(err.msg) return }) // `compile error`
println(g_invalid(9) or { println(err.msg) return }) // `compile error`
println(g_invalid(12) or { println(err.msg) return }) // `compile error`
println(g_invalid(17) or { println(err) return }) // `compile error`
```

**Comment by dumblob** [comment]
This feels really convoluted syntax-wise. Semantics-wise it seems very close or even identical to what is proposed in https://github.com/vlang/rfcs/pull/7 . I see no reason for such a special syntax if under the hood it'll be implemented anyway as described in https://github.com/vlang/rfcs/pull/7 .

Could you elaborate why this special (and arguably unnecessary) syntax should be adopted?

As for the syntax in https://github.com/vlang/rfcs/pull/7 I find it very consise, clear and transparent about the fact, that the return value is a composite type with **equally** important (potentially overlapping) fields - e.g. `int`, `none` and `IError`. Your syntax being irregular seems to imply that the fields are **unequal** which I for myself definitely not want to accept.

**Reply by igotfr** [reply]
explicitly informs what is an error and what is a value with `?`.

It also informs the type of error that can be returned, restricting the domain. In [create_handling.md](https://github.com/vlang/rfcs/pull/7/files?short_path=c0bcf42#diff-c0bcf42c3c53c86a7c34552d5adb11ef246b3a068e77788590d5eaadb952aeb6) only a generic Error type is informed, I like the idea of explicitly informing, as proposed, even if generically, but it is also good to be able to restrict

**Reply by igotfr** [reply]
really I don't care about to separate error and value with `?` or `!` or `|`, but to restrict the error domain I see as necessary, that's what Either does in Haskell and Result in Rust

**Comment by danieldaeschle** [comment]
What would the syntax be for `Error | T` without `none`?

**Reply by igotfr** [reply]
yes, but explicitly informs what is an error and what is a value with `?`.

It also informs the type of error that can be returned, restricting the domain. In [create_handling.md](https://github.com/vlang/rfcs/pull/7/files?short_path=c0bcf42#diff-c0bcf42c3c53c86a7c34552d5adb11ef246b3a068e77788590d5eaadb952aeb6) only a generic Error type is informed, I like the idea of explicitly informing, as proposed, even if generically, but it is also good to be able to restrict

**Reply by igotfr** [reply]
really I don't care about to separate error and value with `?` or `!` or `|`, but to restrict the error domain I see as necessary, that's what Either does in Haskell and Result in Rust

**Reply by igotfr** [reply]
`none` isn't an Error (Result/Either), it's an Option (Maybe)
```v
fn f(x int) (MyError1 | MyError3)?int {}
// or
fn f(x int) Error?int {}
```

**Reply by dumblob** [reply]
I think I understand a bit better what you're aiming for.

But I'm convinced https://github.com/vlang/rfcs/pull/7 is no different in this regard than your proposal.

In https://github.com/vlang/rfcs/pull/7 there is no restriction on a set of "structs" one can return in a sum type. You can just as easily have your `MyError1` `MyError2` `MyError3` and return them (and you actually do not need any casts to `IError` I'd say). Why would you want to cast these to `IError` in the definition of a function? You can cast at the *receiver side* (and actually this might even be casted for you automatically because *V* knows your `MyError...` structs are compatible with `IError` and should let you pass these structs to any function requiring an `IError`).

But maybe I'm missing something (I also didn't manage to follow *V* development for several months now :cry:).

**Reply by igotfr** [reply]
I don't like that errors have an interface as `IError` because it restrict the fields that the `struct` can hold, the only language I know that does this is V, Errors are just structs. The idea of using `?` is explicitly specify what's `Error` and what's value, so the `or` knows what's what, there's nothing special about it, as well as I can pass an `int` as Error in Either, this is considered an error only because it is on the left side, just like Error?int in V

https://github.com/vlang/rfcs/issues/18
https://github.com/vlang/v/discussions/12151

**Reply by igotfr** [reply]
about to restrict error domain, as I said, it's used in Haskell, Rust and many others. and helps to avoid problems, reduce testing and error handling, as it knows what to expect

something similar about domain restriction happens in Dependently Typed and Type-Driven, in languages like Idris, Agda and Ada, which gives a greater level of security for the same reasons: helps to avoid problems, reduce testing and error handling, as it knows what to expect, but this one is for all types and not just errors

**Reply by dumblob** [reply]
I think you're approaching the idea from a bit wrong angle :wink:. Your example would IMHO work:

You can write this:
```v
struct MyError1 {
  msg string
  code int
}
struct MyError3 {
  msg string
  code int
}
fn f( x int ) !int {  // note I use the "accepted syntax" not yours :)
  if x < 5 {
    return 8
  } else if x < 10 {
    return MyError1{ msg: 'Error' }  // no IError casting should be necessary actually
  }
  return MyError3{ msg: 'Invalid' }  // no IError casting should be necessary actually
}

println( f( 4 ) or { println( err.msg ) return } )  // 8
println( f( 9 ) or { println( err.msg ) return } )  // Error
println( f( 12 ) or { println( err.msg ) return } )  // Invalid
```
(@danieldaeschle @JalonSolov could you please confirm the "no IError casting should be necessary actually"?)

But I get that you're aiming for something more complicated - i.e. you want to distinguish which type of error it was. In that case you'd simply use `match` (i.e. dynamic casting):
```v
println( match f( 4 ) {
  int { it }
  MyError1 { println( "meow " + it.msg ) return }
  else { println( "woof " + it.msg ) return }
} )
```

**Reply by danieldaeschle** [reply]
You can pass everything as an interface that fits into it. If MyError1 implements `IError` it could be returned as an error. But currently I'm not sure what happens if you want to return it as an value (and it just have the same interface as IError). This knows @spaceface777

**Reply by igotfr** [reply]
as to whether to use `?` or `!`, I'm indifferent

I liked the idea of not using `IError`

I also liked the `match` as an error/optional handling, but I prefer a differentiation of value and error, like this:
```v
println( match f(4) {
  int { it }
  MyError1 { println( "meow " + err.msg ) return }
  else { println( "woof " + err.msg ) return }
} )
```
and I keep finding important error domain restriction:
```v
fn f( x int ) !int { } // implicitly any Error | None
fn f( x int ) Error!int { } // only any Error but not None
fn f( x int ) None!int { } // only None
fn f( x int ) MyError1 | MyError2 !int { } // explicitly informs the errors
```

**Reply by dumblob** [reply]
> ```v
> fn f( x int ) !int { } // implicitly any Error | None
> fn f( x int ) Error!int { } // only any Error but not None
> fn f( x int ) None!int { } // only None
> fn f( x int ) MyError1 | MyError2 !int { } // explicitly informs the errors
> ```

The semantics you're describing in comments is definitely supported by https://github.com/vlang/rfcs/pull/7 but with the following syntax:

(the only exception is your wish of *implicitly* having both an *error* and *optional* - that's exactly the problem the PR is fighting against and **requires** one to make this combination *explicit*)

```v
fn f( x int )( error | none | int ){ } // explicitly any Error | None
fn f( x int ) !int { } // only any Error but not None
fn f( x int ) ?int { } // only None
fn f( x int )( MyError1 | MyError2 | int ){ } // explicitly informs the errors
```

Does that make more sense now @cindRoberta ?

**Reply by igotfr** [reply]
@dumblob @danieldaeschle these are ok:
```v
fn f( x int ) !int { } // only any Error but not None
fn f( x int ) ?int { } // only None
```
but, in these cases:
```v
fn f( x int )( error | none | int ){ } // explicitly any Error | None
fn f( x int )( MyError1 | MyError2 | int ){ } // explicitly informs the errors
```
how will the compiler distinguish between `error` and value in error handling? for example in `or`:
```v
fn f( x int )( error | none | int ){ } // explicitly any Error | None
f(2) or { err.msg }
```
and in case the function returns a struct, how will it know what the return `struct` and the error `struct`? :
```v
fn f( x int )( MyError1 | MyError2 | MyStruct ){ } // explicitly informs the errors
```
sorry but I don't know how to show my point more clearly. I'll update my request

**Reply by dumblob** [reply]
> but, in these cases:
> 
> ```v
> fn f( x int )( error | none | int ){ } // explicitly any Error | None
> fn f( x int )( MyError1 | MyError2 | int ){ } // explicitly informs the errors
> ```
> 
> how will the compiler distinguish between `error` and value in error handling? for example in `or`:
> 
> ```v
> fn f( x int )( error | none | int ){ } // explicitly any Error | None
> f(2) or { err.msg }
> ```

That's easy. The compiler simply **requires** one to use `match` (i.e. no `or` syntax is allowed). That's the whole point of https://github.com/vlang/rfcs/pull/7 - to make it **explicit** that there is a combination of `error` and `none`.

> and in case the function returns a struct, how will it know what the return `struct` and the error `struct`? :
> 
> ```v
> fn f( x int )( MyError1 | MyError2 | MyStruct ){ } // explicitly informs the errors
> ```

Because sum types are dynamically resolved **in runtime**, this is no problem (it's identical to any other sum type - i.e. you use `match` or `if x is MyError1`). Safety-wise it'd be good to check it in compile time (in this specific case it shall be possible to do).

Btw. this is exactly the question *could you please confirm the "no IError casting should be necessary actually"?* I had above in https://github.com/vlang/v/discussions/12740#discussioncomment-1763298 on @danieldaeschle who replied that @spaceface777 should know the answer. I for myself hope it's planned for *V 1.0* to have this functionality as IMHO it doesn't harm understanding of the source at all but actually improves it. Yet this is only a syntax question and even if it won't be implemented, the behavior is identical with or without (again, because it's a plain old sum type\*).

Feel free to read the RFC https://github.com/vlang/rfcs/pull/7 again and very thoroughly - it's full of wisdom and experience (thanks again @danieldaeschle for putting it together!).

----
\* I should note that under the hood there is plenty of room for optimizations - e.g. `none | int` shouldn't be implemented as a generic sum type but rather similar to `?int` as it's implemented in the current code base.

**Reply by igotfr** [reply]
@dumblob I got it, it's a type checker with `match` not a Error Handling, really my proposal is different from https://github.com/vlang/rfcs/pull/7 but serve the same purpose and there must be a decision on which to implement or both

But certainly a change should be done, because the current Error Handling of V is horrible for a general purpose language and look that I have some experience in syntaxes and semantics for this

**Comment by JalonSolov** [comment]
`none` is a perfectly valid result a lot of times.  For example, an SQL query that doesn't find any matching rows.  Many other languages solve this by always returning an array of rows... which might be an empty array.  Then the caller has to do `if arr.len < 1` if they want to check for that case.  :-\

**Reply by spytheman** [reply]
An empty array is the proper way to represent empty sets; representing them with `none` creates a discontinuity.
To make it more obvious, imagine if instead of `5 - 5 = 0`, we invented a separate word, so `5 - 5 = none`, and then you had to check for that `none` constantly after every arithmetic operation...

**Reply by dumblob** [reply]
I think @JalonSolov knows and agrees to what you write @spytheman . He just used different wording and an example where `... | none` is actually useful to maintain the SQL standard semantics (same goes for trees etc.). Of course, there'll be libraries of both types (those preferring empty arrays and those preferring the SQL semantics). Nobody would want to have `int` to be a disguised `i32 | none` (or `?i32` using the preferred syntax).

Note though, that preferring the SQL semantics has one advantage over empty arrays - it'll be safe. Because V ensures all `... | none` values are explicitly handled. This is not true for for accessing arrays out of bounds - which will compile (and then panic in run time).

**Comment by igotfr** [comment]
related:
https://github.com/vlang/rfcs/issues/17
https://github.com/vlang/rfcs/pull/7
https://github.com/vlang/v/discussions/12151
